fvm/evm/emulator/database/storable.go

```diff
@@ -0,0 +1,434 @@
+		tagNumber, err := dec.DecodeTagNumber()
+		if err != nil {
+			return nil, err
+		}
+
+		switch tagNumber {
+
+		case atree.CBORTagStorageID:
+			return atree.DecodeStorageIDStorable(dec)
+
+		case cborTagUInt8Value:
+			n, err := dec.DecodeUint64()
+			if err != nil {
+				return nil, err
+			}
+			if n > math.MaxUint8 {
+				return nil, fmt.Errorf("invalid data, got %d, expected max %d", n, math.MaxUint8)
+			}
+			return Uint8Value(n), nil
+
+		case cborTagUInt64Value:
+			n, err := dec.DecodeUint64()
+			if err != nil {
+				return nil, err
+			}
+			return Uint64Value(n), nil
+
+		default:
+			return nil, fmt.Errorf("invalid tag number %d", tagNumber)
+		}
+
+	default:
+		return nil, fmt.Errorf("invalid cbor type %s for storable", t)
+	}
+}
+
+func compare(storage atree.SlabStorage, value atree.Value, storable atree.Storable) (bool, error) {
+	switch v := value.(type) {
+
+	case Uint8Value:
+		other, ok := storable.(Uint8Value)
+		if !ok {
+			return false, nil
+		}
+		return uint8(other) == uint8(v), nil
+
+	case Uint64Value:
+		other, ok := storable.(Uint64Value)
+		if !ok {
+			return false, nil
+		}
+		return uint64(other) == uint64(v), nil
+
+	case ByteStringValue:
+		other, ok := storable.(ByteStringValue)
+		if ok {
+			return bytes.Equal(other.data, v.data), nil
+		}
+
+		// Retrieve value from storage
+		otherValue, err := storable.StoredValue(storage)
+		if err != nil {
+			return false, err
+		}
+		other, ok = otherValue.(ByteStringValue)
+		if ok {
+			return bytes.Equal(other.data, v.data), nil
+		}
+
+		return false, nil
+	}
+
+	return false, fmt.Errorf("value %T not supported for comparison", value)
+}
+
+func hashInputProvider(value atree.Value, buffer []byte) ([]byte, error) {
+	switch v := value.(type) {
+	case Uint8Value:
+		return v.getHashInput(buffer)
+	case Uint64Value:
+		return v.getHashInput(buffer)
+	case ByteStringValue:
+		return v.getHashInput(buffer)
+	}
+
+	return nil, fmt.Errorf("value %T not supported for hash input", value)
+}
+
+type testTypeInfo struct {
+	value uint64
+}
+
+var _ atree.TypeInfo = testTypeInfo{}
+
+func (i testTypeInfo) Encode(e *cbor.StreamEncoder) error {
+	return e.EncodeUint64(i.value)
+}
+
+func (i testTypeInfo) Equal(other atree.TypeInfo) bool {
+	otherTestTypeInfo, ok := other.(testTypeInfo)
+	return ok && i.value == otherTestTypeInfo.value
+}
+
+func decodeTypeInfo(dec *cbor.StreamDecoder) (atree.TypeInfo, error) {
+	value, err := dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+
+	return testTypeInfo{value: value}, nil
+}
+
+func NewPersistentSlabStorage(baseStorage atree.BaseStorage) (*atree.PersistentSlabStorage, error) {
+
+	encMode, err := cbor.EncOptions{}.EncMode()
+	if err != nil {
+		return nil, err
+	}
+
+	decMode, err := cbor.DecOptions{}.DecMode()
+	if err != nil {
+		return nil, err
+	}
+
+	return atree.NewPersistentSlabStorage(
+		baseStorage,
+		encMode,
+		decMode,
+		decodeStorable,
+		decodeTypeInfo,
+	), nil
+
+}
+
+type typeInfo struct{}
```

Comment:
Maybe rename this to `emptyTypeInfo`, like Cadence storage map.

```suggestion
type emptyTypeInfo struct{}
```

Reply:
renamed.